src/flamenco/txn/fd_txn_generate.c

```diff
@@ -0,0 +1,160 @@
+  txn_meta->recent_blockhash_off = (ushort)(txn_meta->acct_addr_off + (txn_meta->acct_addr_cnt * FD_TXN_ACCT_ADDR_SZ));
+
+  /* Fill message header in txn payload */
+  uchar * write_ptr = out_txn_payload + txn_meta->message_off;
+  fd_txn_message_hdr_t msg_header = { .num_signatures = accounts->signature_cnt,
+                                      .num_readonly_signatures = accounts->readonly_signed_cnt,
+                                      .num_readonly_unsigned   = accounts->readonly_unsigned_cnt };
+  memcpy( write_ptr, &msg_header, sizeof(fd_txn_message_hdr_t) ) ;
+  write_ptr += sizeof(fd_txn_message_hdr_t);
+
+  /* Write number of accounts */
+  *write_ptr = (uchar)txn_meta->acct_addr_cnt;
+  write_ptr += 1;
+
+  /* Write accounts list to txn payload */
+  ulong signers_write_sz = FD_TXN_ACCT_ADDR_SZ * (accounts->signature_cnt - accounts->readonly_signed_cnt);
+  fd_memcpy( write_ptr, accounts->signers_w, signers_write_sz );
+  write_ptr += signers_write_sz;
+
+  fd_memcpy( write_ptr, accounts->signers_r, FD_TXN_ACCT_ADDR_SZ * accounts->readonly_signed_cnt );
+  write_ptr += FD_TXN_ACCT_ADDR_SZ * accounts->readonly_signed_cnt;
+
+  ulong non_signers_write_sz = FD_TXN_ACCT_ADDR_SZ * (ulong)(accounts->acct_cnt - accounts->readonly_unsigned_cnt - accounts->signature_cnt);
+  fd_memcpy( write_ptr, accounts->non_signers_w, non_signers_write_sz);
+  write_ptr += non_signers_write_sz;
+
+  fd_memcpy( write_ptr, accounts->non_signers_r, FD_TXN_ACCT_ADDR_SZ * accounts->readonly_unsigned_cnt );
+  write_ptr += FD_TXN_ACCT_ADDR_SZ * accounts->readonly_unsigned_cnt;
+  FD_TEST( (ushort)((ulong)write_ptr - (ulong)out_txn_payload) == txn_meta->recent_blockhash_off );
+
+  /* Write recent blockhash */
+  if( FD_LIKELY( opt_recent_blockhash ) ) {
+    memcpy( write_ptr, opt_recent_blockhash, FD_TXN_BLOCKHASH_SZ );
+  } else {
+    memset( write_ptr, 0UL, FD_TXN_BLOCKHASH_SZ );
+  }
+  write_ptr += FD_TXN_BLOCKHASH_SZ;
+
+  return (ulong)(write_ptr - out_txn_payload);
+}
+
+ulong
+fd_txn_add_instr( uchar * txn_meta_ptr,
+                  uchar out_txn_payload[ static FD_TXN_MTU ],
+                  uchar program_id,
+                  uchar const * accounts,
+                  ulong accounts_sz,
+                  uchar const * instr_buf,
+                  ulong instr_buf_sz ) {
+
+  fd_txn_t * txn_meta = (fd_txn_t *) txn_meta_ptr;
+  FD_TEST( txn_meta->instr_cnt < FD_TXN_INSTR_MAX );
+  FD_TEST( txn_meta->recent_blockhash_off != 0 );
+
+  uchar * instr_start = out_txn_payload + txn_meta->recent_blockhash_off + FD_TXN_BLOCKHASH_SZ;
+  txn_meta->instr_cnt++;
+  uchar * write_ptr       = instr_start;
+
+  uint compact_instr_cnt_sz = fd_cu16_enc( (ushort)txn_meta->instr_cnt, write_ptr );
+  FD_TEST( compact_instr_cnt_sz == 1 );
+
+  write_ptr += compact_instr_cnt_sz;
+
+  /* Calculate offset of next instruction. */
+  if ( FD_UNLIKELY( txn_meta->instr_cnt > 1 ) ) {
```

Comment:
I am not sure why this is `txn_meta->instr_cnt > 1` instead of `txn_meta->instr_cnt > 0`. It seems that, even if `txn_meta->instr_cnt == 1`, we need to recalculate `write_ptr`.

Reply:
the write_ptr is already at the right offset for that case